---
title: "Adult vs Fetal eye tx"
output: html_notebook
---



```{r setup}

knitr::opts_knit$set(root.dir = '~/NIH/occular_transcriptomes_paper/')
```



```{r}
library(tidyverse)
load('/Volumes/data/occular_transcriptomes_paper/clean_data/exon_classification.Rdata')
gtf <- rtracklayer::readGFF('all_tissues.combined.gtf')
tcon_tab <- read_tsv('gfc_TCONS_to_st_MSTRG.tsv')
eye_tissues <- colnames(tcon_tab)[2:7] # might change, dbl check
tcon_tab_eye <- tcon_tab %>% select(transcript_id, eye_tissues) %>% filter( (apply(.[,-1],1, function(x) sum(is.na(x))))<6)
gtf_eye <- filter(gtf, transcript_id %in% tcon_tab_eye$transcript_id)

compare_exon_counts <- function(tissue){ #count the number of distinct expressed exons per gene fetal and Adult
    
    ages  <- paste0(tissue, c('_Adult.Tissue', '_Fetal.Tissue'))
    
    count_exons <- function(cond){
        k <- ifelse(grepl('Adult', cond), 'Adult', 'Fetal')
        cond_tx<-  tcon_tab_eye %>% select(transcript_id, cond) %>% filter(!is.na(.[,2])) %>% pull(transcript_id)
        gtf %>% filter(transcript_id %in% cond_tx, type == 'exon') %>% 
        select(seqid, strand, start, end, gene_name) %>% 
        group_by(gene_name) %>% summarise(exon_count=n()) %>% pull(exon_count)
            #mutate(dev=k, gene_name=NULL)
    }
    
    res <- lapply(ages, count_exons)
    names(res) <- c('Adult', 'Fetal')
    
    # ggplot(data=(res %>% filter(exon_count<=100))) + 
    #      geom_violin(aes(y=exon_count, x=dev))+
    #     coord_flip() +
    #     theme_minimal()

    #print(sapply(res, length))
    test <-  c(t.test(res$Adult, res$Fetal, alternative = 'greater')[['p.value']],  
              t.test(res$Adult, res$Fetal, alternative = 'less')[['p.value']]) 
    names(test) <- c('greater', 'less')
    sig <- which.min(test)
    return(test[sig])
    
}
#Adult  vs Fetal
compare_exon_counts('Retina')
compare_exon_counts('RPE')
compare_exon_counts('Cornea')

```


```{r}
library(ggpubr)
compare_exon_counts_and_plot <- function(tissue){ #count the number of distinct expressed exons per gene fetal and Adult
    
    ages  <- paste0(tissue, c('_Adult.Tissue', '_Fetal.Tissue'))
    
    count_exons <- function(cond){
        k <- ifelse(grepl('Adult', cond), 'Adult', 'Fetal')
        cond_tx<-  tcon_tab_eye %>% select(transcript_id, cond) %>% filter(!is.na(.[,2])) %>% pull(transcript_id)
        gtf %>% filter(transcript_id %in% cond_tx, type == 'exon') %>% 
        select(seqid, strand, start, end, gene_name) %>% 
        group_by(gene_name) %>% summarise(exon_count=n()) %>%
            mutate(dev=k, gene_name=NULL)
    }
    
    res <- lapply(ages, count_exons) %>% bind_rows()
    
   plot <-  ggboxplot(res, x='dev', y='exon_count', 
              title = paste0('comparison of # Distinct exons per genes in ', tissue),
              xlab = 'Developmental point')+
        stat_compare_means(method = 't.test',)+
        #coord_flip() +
        theme_minimal()
    print(plot)
    
}

compare_exon_counts_and_plot('Retina')
compare_exon_counts_and_plot('RPE')
compare_exon_counts_and_plot('Cornea')


```

```{r}
library(ggpubr)
compare_exon_counts_and_plot <- function(tissue){ #count the number of distinct expressed exons per gene fetal and Adult
    
    ages  <-  c('Retina_Adult.Tissue', 'RPE_Adult.Tissue')
    
    count_exons <- function(cond){
        k <- ifelse(grepl('Retina', cond), 'Retina', 'RPE')
        cond_tx<-  tcon_tab_eye %>% select(transcript_id, cond) %>% filter(!is.na(.[,2])) %>% pull(transcript_id)
        gtf %>% filter(transcript_id %in% cond_tx, type == 'exon') %>% 
        select(seqid, strand, start, end, gene_name) %>% 
        group_by(gene_name) %>% summarise(exon_count=n()) %>%
            mutate(dev=k, gene_name=NULL)
    }
    
   res <- lapply(ages, count_exons) %>% bind_rows()
    
   plot <-  ggboxplot(res, x='dev', y='exon_count', 
              title = paste0('comparison of # Distinct exons per genes in ', tissue),
              xlab = 'Developmental point')+
        stat_compare_means(method = 't.test',)+
        #coord_flip() +
        theme_minimal()
    print(plot)
    
}




```

what do I want to show? Genes specific to retinal biology have a noncoding/alternate minor isoform in fetal retina


```{r}
library(tidyverse)
library(enrichR)
library(UpSetR)
library(matrixStats)
setwd('~/NIH/occular_transcriptomes_paper/')
load('clean_data/V1_exon_classification_data.Rdata')
gtf <- rtracklayer::readGFF('all_tissues.combined_NovelAno.gtf')
anno_tab <- gtf %>% filter(type == "transcript") %>% select(transcript_id, gene_name, oId)
tc2ms <- read_tsv('gfc_TCONS_to_st_MSTRG.tsv') %>% filter(!transcript_id %in% novel_loci_distinct$transcript_id)
eye_tissues <- c('Retina_Fetal.Tissue', 'Retina_Adult.Tissue', 'RPE_Fetal.Tissue', 'RPE_Adult.Tissue', 'Cornea_Adult.Tissue',
                 'Cornea_Fetal.Tissue', "ESC_StemCellLine")
sample_table_eye <- read_tsv('~/NIH/eyeintegration_splicing/sampleTableV6.tsv') %>% 
    filter(subtissue %in% eye_tissues)

retnet <- scan('/Volumes/data/eyeintegration_splicing/ref/retnet_hgncIDs_2017-03-28.txt', '\n')
#this is our baseline sets, all transcripts that were only built in the eye 
t2g <- gtf %>% filter(type == 'transcript') %>% select(transcript_id, gene_name) %>% distinct 
load('all_tissues_complete_quant.rdata') 
complete_quant[is.na(complete_quant)] <- 0
counts_eye <- complete_quant[,c('transcript_id', sample_table_eye$sample)]
counts_eye_by_tissue <- lapply(eye_tissues, 
                               function(tis) filter(sample_table_eye, subtissue == tis) %>% pull(sample) %>%
                                   {counts_eye[,c('transcript_id', .)]} %>%
                                   mutate(!!tis := rowMedians(.[,-1] %>% as.matrix)) %>%
                                   select(transcript_id, !!tis) ) %>% reduce(left_join) %>% left_join(t2g, .)

keep <- rowSums(counts_eye_by_tissue[,-(1:2)]) > 0
med_0_counts_eye <- counts_eye_by_tissue[!keep,]
counts_eye_by_tissue <- counts_eye_by_tissue[keep,]



calc_isoform_percentage <- function(t_tissue){
    df <- counts_eye_by_tissue %>% select(transcript_id, gene_name, !!t_tissue)
    tt_sym <- as.symbol(t_tissue)
    df_gene_sums <- df %>% 
        select(-transcript_id) %>% 
        group_by(gene_name) %>%  
        summarise(gene_sums:= sum(!!tt_sym)) %>% 
        left_join(df, .) %>% 
        mutate(piu = .[[t_tissue]] / .[['gene_sums']], !!t_tissue :=NULL ) %>% 
        select(transcript_id, gene_name, !!t_tissue:=piu)
    return(df_gene_sums)
    
}

piu_raw <- lapply(colnames(counts_eye_by_tissue)[-(1:2)], calc_isoform_percentage) %>% reduce(left_join)
source('~/scripts/df_replace_nan.R')
piu <-replace_nan(piu_raw)


```



now only retina fetal 

```{r}
retina_fetal_spec_tx <- retina_spec_det %>% filter(Retina_Fetal.Tissue, !Retina_Adult.Tissue) %>% 
    pull(transcript_id) %>% unique()
retina_fetal_spec_gene <- filter(gtf, transcript_id %in% retina_fetal_spec_tx ) %>% pull(gene_name) %>% unique

```

Ok. Now, lets look what whats making these transcripts novel


```{r}
gtf_ano <- rtracklayer::readGFF('all_tissues.combined_NovelAno.gtf')
rf_spec_tx_gtf <- filter(gtf_ano, transcript_id %in% retina_fetal_spec_tx, !grepl('TCON', gene_name))
rf_spec_tx_gtf %>% select(novel_exon_id, novel_exon_type) %>% filter(!is.na(novel_exon_id)) %>% distinct %>% pull(novel_exon_type) %>% table 


rf_spec_tx_gtf %>% filter(type  == 'transcript') %>% pull(transcript_type) %>% table 
rf_spec_tx_gtf %>% filter(transcript_type == 'protein_coding') %>% 
    select(novel_exon_id, exon_type, novel_exon_type) %>%filter(!is.na(novel_exon_id)) %>% distinct %>% pull(exon_type) %>% table 
filter(rf_spec_tx_gtf, gene_name %in% retnet, transcript_type != 'protein_coding') %>% select(novel_exon_id)

keep_codes <- 
```

```{r}
novelfilt <- function(x){x %>%  select(novel_exon_id, exon_type, novel_exon_type) %>%filter(!is.na(novel_exon_id)) %>% distinct}
rf_retnet <- filter(rf_spec_tx_gtf, gene_name %in% retnet) 
rf_retnet %>% filter(type == 'transcript') %>% pull(transcript_type) %>% table 

rf_retnet %>% filter(transcript_type == 'protein_coding') %>% novelfilt %>% pull(novel_exon_type) %>% table 
rf_retnet %>% filter(transcript_type != 'protein_coding') %>% novelfilt %>% pull(novel_exon_type) %>% table 
rf_retnet %>% filter(transcript_type == 'protein_coding') %>% novelfilt %>% pull(exon_type) %>% table 
rf_retnet %>% filter(transcript_type != 'protein_coding') %>% novelfilt %>% pull(exon_type) %>% table 

```

