---
title: "R Notebook"
output: html_notebook
---

```{r setup}
knitr::opts_knit$set(root.dir = '~/NIH/occular_transcriptomes_paper/')
```



```{r}
library(tidyverse)
library(enrichR)
library(UpSetR)
library(matrixStats)
library(ggpubr)
setwd('~/NIH/occular_transcriptomes_paper/')
load('clean_data/V1_exon_classification_data.Rdata')
gtf <- rtracklayer::readGFF('all_tissues.combined_NovelAno.gtf')
anno_tab <- gtf %>% filter(type == "transcript") %>% select(transcript_id, gene_name, oId)
tc2ms <- read_tsv('gfc_TCONS_to_st_MSTRG.tsv') %>% filter(!transcript_id %in% novel_loci_distinct$transcript_id)
eye_tissues <- c('Retina_Fetal.Tissue', 'Retina_Adult.Tissue', 'RPE_Fetal.Tissue', 'RPE_Adult.Tissue',
                 'Cornea_Adult.Tissue', 'Cornea_Fetal.Tissue', "ESC_StemCellLine")
sample_table_eye <- read_tsv('~/NIH/eyeintegration_splicing/sampleTableV6.tsv') %>% 
    filter(subtissue %in% eye_tissues)

retnet <- scan('/Volumes/data/eyeintegration_splicing/ref/retnet_hgncIDs_2017-03-28.txt', '\n')
#this is our baseline sets, all transcripts that were only built in the eye 
t2g <- gtf %>% filter(type == 'transcript') %>% select(transcript_id, gene_name) %>% distinct 
load('all_tissues_complete_quant.rdata') 
complete_quant[is.na(complete_quant)] <- 0
counts_eye <- complete_quant[,c('transcript_id', sample_table_eye$sample)]
counts_eye_by_tissue <- lapply(eye_tissues, 
                               function(tis) filter(sample_table_eye, subtissue == tis) %>% pull(sample) %>%
                                   {counts_eye[,c('transcript_id', .)]} %>%
                                   mutate(!!tis := rowMedians(.[,-1] %>% as.matrix)) %>%
                                   select(transcript_id, !!tis) ) %>% reduce(left_join) %>% left_join(t2g, .)

keep <- rowSums(counts_eye_by_tissue[,-(1:2)]) > 0
med_0_counts_eye <- counts_eye_by_tissue[!keep,]
counts_eye_by_tissue <- counts_eye_by_tissue[keep,]



calc_isoform_percentage <- function(t_tissue){
    df <- counts_eye_by_tissue %>% select(transcript_id, gene_name, !!t_tissue)
    tt_sym <- as.symbol(t_tissue)
    df_gene_sums <- df %>% 
        select(-transcript_id) %>% 
        group_by(gene_name) %>%  
        summarise(gene_sums:= sum(!!tt_sym)) %>% 
        left_join(df, .) %>% 
        mutate(piu = .[[t_tissue]] / .[['gene_sums']], !!t_tissue :=NULL ) %>% 
        select(transcript_id, gene_name, !!t_tissue:=piu)
    return(df_gene_sums)
    
}

piu_raw <- lapply(colnames(counts_eye_by_tissue)[-(1:2)], calc_isoform_percentage) %>% reduce(left_join)
source('~/scripts/df_replace_nan.R')
piu <-replace_nan(piu_raw)

det_mat <- apply(tc2ms[,-1],2, function(t) !is.na(t)) %>% cbind(tc2ms[,1], .)
det_eye <- det_mat[,eye_tissues] %>% {rowSums(.) > 0}
not_det_body <- det_mat %>% as.data.frame %>% 
    select(-eye_tissues, -transcript_id, -Lens_Stem.Cell.Line) %>% {rowSums(.) == 0}
eye_spec_tx_det <- filter(det_mat, det_eye, not_det_body) %>% select(c('transcript_id',eye_tissues))
ESC_only_tx <-  filter(eye_spec_tx_det, ESC_StemCellLine, 
                       rowSums(eye_spec_tx_det[,-1] %>% as.data.frame() %>% select(-ESC_StemCellLine)) == 0) %>% 
    pull(transcript_id)  
eye_spec_tx <- filter(eye_spec_tx_det, !transcript_id %in% ESC_only_tx, 
                                       !transcript_id %in% novel_loci_distinct$transcript_id, 
                                        transcript_id %in% counts_eye_by_tissue$transcript_id) %>% 
    select(-ESC_StemCellLine)
eye_spec_tx %>% pull(transcript_id) %>% unique() %>% length 

eye_spec_genes <- filter(gtf, transcript_id %in% eye_spec_tx$transcript_id) %>% pull(gene_name) %>% unique


```

Fig - comparison of trasncript contruction of novel vs ref tx across eye

```{r}
determine_specifcity <- function(x,df){
    #nspec_o <- df %>% select(-!!x, -transcript_id) %>% {rowSums(.) == 0}
    df %>% filter(.[,x]) %>% pull(transcript_id)
}
et <- eye_tissues <- c('Retina_Fetal.Tissue', 'Retina_Adult.Tissue', 'RPE_Fetal.Tissue', 'RPE_Adult.Tissue',
                 'Cornea_Adult.Tissue', 'Cornea_Fetal.Tissue')

novel_eyespec <- filter(eye_spec_tx, transcript_id %in% novel_transcripts$transcript_id)
ref_eyespec <- filter(eye_spec_tx, !transcript_id %in% novel_transcripts$transcript_id)

novel_eye_tx_by_tissue <- lapply(et, determine_specifcity, novel_eyespec)
names(novel_eye_tx_by_tissue) <- et


ref_eye_tx_by_tissue <- lapply(et, determine_specifcity, ref_eyespec)
names(ref_eye_tx_by_tissue) <- et
```


```{r}
```


```{r}
upset(fromList(novel_eye_tx_by_tissue),sets = et, order.by = 'freq',keep.order = T  )
upset(fromList(ref_eye_tx_by_tissue),sets = et ,order.by = 'freq', keep.order = T)
```


```{r}
```



Lets narrow our search field. Focus only on novel_retina
```{r}
summarise_novel_transcripts_in_tissue <- function(t_tissue) {
    eye_spec_novel <- filter(eye_spec_tx, transcript_id %in% novel_transcripts$transcript_id)
    fetal <- paste0(t_tissue, '_Fetal.Tissue')
    adult <- paste0(t_tissue, '_Adult.Tissue')
    exp_in_tissue <- (eye_spec_novel[,-1] %>% select(contains(t_tissue)) %>% {rowSums(.) >0})
    not_exp_othertissues <-  eye_spec_novel[,-1] %>% select(-contains(t_tissue)) %>% {rowSums(.) == 0}
    tissue_spec_det <- eye_spec_novel[exp_in_tissue & not_exp_othertissues, ] 
    
    tissue_spec_det <- tissue_spec_det[,c('transcript_id', fetal, adult)]
    tissue_spec_det %>% pull(transcript_id) %>% unique() %>% length 
    tissue_spec_genes <-  tissue_spec_det %>% pull(transcript_id) %>% unique() %>% 
        {filter(gtf, transcript_id %in% .)} %>% pull(gene_name) %>% unique()
    fetal_exp <- tissue_spec_det %>% filter(.[,fetal]) 
    adult_exp <- tissue_spec_det %>% filter(.[,adult]) 
    both <- tissue_spec_det %>% filter(.[,adult], .[,fetal])
    # 
    # adult_piu_ns <- det_mat %>% 
    #     filter(.[,adult]) %>% 
    #     pull(transcript_id) %>% 
    #     {filter(piu, transcript_id %in% ., 
    #                  !transcript_id %in% tissue_spec_det$transcript_id,
    #                  !transcript_id %in% novel_transcripts$transcript_id ) } %>% 
    #     select(transcript_id, piu := !!adult) %>% 
    #     mutate(tissue=t_tissue, stage='adult_ns')
    # 
    # fetal_piu_ns <- det_mat %>% 
    #     filter(.[,fetal]) %>% 
    #     pull(transcript_id) %>% 
    #     {filter(piu, transcript_id %in% ., 
    #                  !transcript_id %in% tissue_spec_det$transcript_id,
    #                  !transcript_id %in% novel_transcripts$transcript_id ) } %>% 
    #     select(transcript_id, piu := !!fetal) %>% 
    #     mutate(tissue=t_tissue, stage='fetal_ns')
    
    
    
    adult_piu_spec <- tissue_spec_det %>% 
        filter(.[,adult]) %>% 
        pull(transcript_id) %>% 
        {filter(piu, transcript_id %in% .)} %>% 
        select(transcript_id, piu := !!adult) %>% 
        mutate(tissue=t_tissue, stage='adult')
    
    fetal_piu_spec <- tissue_spec_det %>% 
        filter(.[,fetal]) %>% 
        pull(transcript_id) %>% 
        {filter(piu, transcript_id %in% .)} %>% 
        select(transcript_id, piu := !!fetal) %>% 
        mutate(tissue=t_tissue, stage='fetal')
    
    
    
    fetal_locations <- gtf %>% 
        filter(transcript_id %in% fetal_exp$transcript_id) %>% 
        select(id, nv_type_rc, exon_location) %>% 
        filter(!is.na(id)) %>% 
        distinct %>% 
        pull(exon_location) %>% 
        table %>% 
        {./sum(.)}
    
    
    adult_locations <- gtf %>% 
        filter(transcript_id %in% adult_exp$transcript_id) %>% 
        select(id, nv_type_rc, exon_location) %>% 
        filter(!is.na(id)) %>% 
        distinct %>% 
        pull(exon_location) %>% 
        table %>% 
        {./sum(.)}
    list(loc_df=bind_rows(adult_locations, fetal_locations) %>% mutate(age=c('adult', 'fetal'), tissue=t_tissue) ,
         piu_df=bind_rows(adult_piu_spec, fetal_piu_spec)
        )
}

tissues <- c('Retina', 'Cornea', 'RPE')
res <- lapply(tissues, summarise_novel_transcripts_in_tissue)


location_df <- lapply(res, function(x) x[['loc_df']]) %>% bind_rows %>% gather(location, count, -age, -tissue)
piu_df <- lapply(res, function(x) x[['piu_df']]) %>% bind_rows

ggplot(piu_df) + 
    geom_boxplot(aes(x=stage, y=piu, color = tissue)) + 
    facet_wrap(~tissue)


p <- ggboxplot(piu_df, x='stage', y='piu', color = 'tissue',
               title ='Comparison of percent isoform usage(piu) of novel \ntranscripts in fetal and adult eye tissues')+
    stat_compare_means(label.y = 1.1) +
    scale_color_manual(values = c('green', 'blue', 'red'))
    facet(p,facet.by = 'tissue')+
    theme_minimal()

# gtf %>% 
#     filter(transcript_id %in% fetal_exp$transcript_id) %>% 
#     select(id, nv_type_rc, exon_location) %>% 
#     filter(!is.na(id)) %>% 
#     distinct %>% 
#     pull(nv_type_rc) %>% 
#     table 
# gtf %>% 
#     filter(transcript_id %in% adult_exp$transcript_id) %>% 
#     select(id, nv_type_rc, exon_location) %>% 
#     filter(!is.na(id)) %>% 
#     distinct %>% 
#     pull(nv_type_rc) %>% 
#     table %>% 
#     {./sum(.)}




```


```{r}
ggplot(location_df) + 
    geom_bar(aes(x=age, fill=location, y=count), position = 'fill', stat = 'identity' ) + 
    facet_wrap(~ tissue) + 
    ylab('percentage of novel exons') + 
    ggtitle('location of novel exons in occular tissues')



```





```{r}
ret_txs <- filter(det_mat, transcript_id %in% novel_transcripts$transcript_id, Retina_Fetal.Tissue|Retina_Adult.Tissue) %>% pull(transcript_id) %>% {filter(gtf, transcript_id %in% .)} %>% pull(gene_name) %>% unique
all_ret_genes <- filter(det_mat, Retina_Fetal.Tissue|Retina_Adult.Tissue) %>% pull(transcript_id) %>% {filter(gtf, transcript_id %in% .)} %>% pull(gene_name) %>% unique
m=length(retnet)#white balls
n=length(all_ret_genes) - m#black balls
k=length(ret_txs)#draws
x=sum(ret_txs %in% retnet)#successes
dhyper(x,m,n,k)
```




```{r}

```

