---
title: '*De novo* transcriptomes built from hundreds of eye tissues reveal hundreds of novel gene isoforms'
author: 'Vinay S Swamy, David McGaughey'
output:
  word_document: 
            reference_docx: reference_doc_v1.docx  
  html_document:
    df_print: paged
---

```{r setup, echo=FALSE}
#knitr::opts_knit$set(root.dir = '~/NIH/occular_transcriptomes_paper/')
knitr::opts_knit$set(root.dir = '/Volumes/data/ocular_transcriptomes_paper/')
#knitr::opts_knit$set(root.dir = '/data/swamyvs/occular_transcriptomes_paper')
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```


```{r prep}
library(tidyverse)
library(DT)
library(RColorBrewer)
library(UpSetR)
library(ComplexHeatmap)
library(ggpubr)
library(knitr)
library(grid)
library(png)
sample_table <- read_tsv('sampleTableFull.tsv') %>% filter(subtissue != 'synth')
load('clean_data/rdata/tissue_to_colors.Rdata')
load('/Volumes/data/ocular_transcriptomes_pipeline/ref/core_tight.Rdata')
tissue_color_mapping_df <- bind_rows(tissue_color_mapping_df,  tibble(body_location=c('Brain(avg)', 'Body(avg)'), color=c('orange','yellow')))
load('clean_data/rdata/transcriptome_pipeline_stats.Rdata')
subtissue_to_bodyloc <- sample_table %>% select(subtissue, body_location) %>% distinct
core_tight <- core_tight %>% select(sample=sample_accession, study=study_accession) %>% distinct
```


# Results 

## A rigorous analysis pipeline finds thousands of novel gene isoforms

```{r}
df <- sample_table %>% 
  left_join(core_tight) %>% 
  filter(!body_location %in% c('Body', 'Brain', 'ESC_Stem.Cell.Line', 'Lens_Stem.Cell.Line')) %>% 
  group_by(subtissue) %>%
  summarise(`number of samples` = n(), `number of studies` = length(unique(study))) %>% 
  arrange(desc(`number of studies`)) %>% 
  inner_join(tx_counts %>% select(subtissue, `transcriptome size`=final)) %>% 
  mutate(subtissue=gsub('_|\\.',' ', subtissue)) 
kable(df)
```

::: {custom-style="CustomCaption"}
Table 1. Statistics about ocular sample dataset. Transcriptome size is defined as the number of unique transcripts expressed in a given tissue type
:::

|        We built transcriptomes using 340 [publicly]{custom-style="CaptionCustom"} available ocular tissue samples. We include both adult and fetal tissue from the Cornea, Retina, Retinal Pigmented Epithelium(RPE) mined from 32 different studies(Table 1) Our fetal tissues consist of both human fetal tissues and human induced pluripotent stem cell(iPSC) derived tissue. We supplemented our ocular data set with 905 samples across 46 body locations from the GTEx project.



![](/Volumes/data/ocular_transcriptomes_pipeline/dag.png)

::: {custom-style="CustomCaption"}
Figure 1. Directed Acyclic Graph of *de novo* transcriptome construction pipeline
We designed a Snakemake pipeline to efficiently and reproducibly process and build our *de novo* transcriptomes(fig1), used it to construct a separate tissue specific transcriptome for each tissue type.
:::

|        After initial construction of transcripts, we found over two million distinct transcripts (sup table 1). We saw that many of these transcripts were detected in one or two samples(sup fig 1) so we refined our set of transcriptomes using series of filtering steps(methods) to remove low confidence transcripts. Our final transcriptome contains 340,456 distinct transcripts with 102,323 previously annotated and 238,677 novel transcripts, and incudes 300 kilobases of previously unannotated genomic sequence. We define novel as any region of genomic sequence that is not found in Gencode, Ensembl and Refseq annotation databases. Novel transcripts are split into two categories: novel isoforms which are novel variations of known genes, and novel loci, which are previously unreported, entirely novel regions of transcribed sequence. 

```{r transcriptome_pipeline_stats, fig.height=13, fig.width=16}
#  novel_loci bar graph, novel isoform bargraphs initial tx count, final tx count
load('clean_data/rdata/buildResultsSummary.Rdata')
novel_transcripts_per_tissue <- novel_transcripts_per_tissue %>% filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% mutate(body_location_pretty=gsub('_|\\.', ' ', body_location))
color_list<- novel_transcripts_per_tissue$color
names(color_list) <- novel_transcripts_per_tissue$body_location
isoforms <- ggplot(data = novel_transcripts_per_tissue) +
    geom_col(aes(x=body_location_pretty, y=novel_transcript_count, fill=body_location)) +
    scale_fill_manual(values = color_list)+
    ggtitle('Novel Isoforms Contructed Across the Body')+
    ylab('number of novel transcripts')+
    xlab('Body Location')+
    theme(axis.text.x=element_text(angle=45, hjust = 1))
novel_loci_per_tissue <- novel_loci_per_tissue %>% filter(transcript_type!='color',!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% mutate(counts=as.numeric(counts), body_location_pretty=gsub('_|\\.', ' ', body_location))

novel_exons <- as_ggplot(rasterGrob(readPNG('clean_data/plots/sankey_novel_exons.png'), interpolate = T))


loci <- ggplot(data = novel_loci_per_tissue) +
    geom_col(aes(x=body_location_pretty, y=counts, fill=body_location, alpha=transcript_type), position = 'dodge') +
    scale_fill_manual(values = color_list) +
    scale_alpha_discrete(range=c(.5,1)) +
    ggtitle('Novel Loci constructed in Occular Tissues', ) +
    ylab('number of novel loci detected') +
    xlab('Body Location\n\n\n') +
    theme(axis.text.x=element_text(angle=45, hjust = 1))

p0 <- ggarrange(isoforms, loci, ncol = 2, labels = c('A', 'B'), font.label = list(size=25))
p1 <- annotate_figure(novel_exons, fig.lab = 'C', fig.lab.pos = 'top.left', fig.lab.size = 25)
ggarrange(p0, p1, nrow=2)
```

::: {custom-style="CustomCaption"}
Figure 2 
A,B - Number of novel gene isoforms and novel loci constructed across ocular and other body tissues.Brain and body represent an average of 13 and 34 distinct subtissues, respectively
C - Classification of novel gene isoforms based on source of novelty 
:::  
  
```{r}
# load('/Volumes/data/ocular_transcriptomes_pipeline/data/rdata/novel_exon_classification.Rdata')
# gtf <- rtracklayer::readGFF('/Volumes/data/ocular_transcriptomes_pipeline/data/gtfs/all_tissues.combined_NovelAno.gtf')
# # 
# # gtf_exons <- gtf %>% filter(type == 'exon')
# omit_exons <- gtf %>% filter(type == 'exon', !grepl('ENST', oId)) %>% group_by(transcript_id) %>% summarise(num_novel_exon=sum(is.na(novel_exon_id))) %>% filter(num_novel_exon == 0) %>% pull(transcript_id)
# event_counts <-  novel_exons_TSES %>% pull(nv_type_rc) %>% table %>% {tibble(type=names(.), count=.)} %>% 
#   mutate(type = case_when( type == 'A3SS'| type == 'A5SS' ~ 'Alternative Splice Site', 
#                            type == 'novel_exon' ~ ' Fully Unannotated Exon',
#                            type == 'RI' ~ 'Retained Intron', 
#                            type == 'novel_TES' ~ 'Novel Transcriptional End Site', 
#                            type == 'novel_TSS' ~ 'Novel Transcriptional Start Site')) %>% group_by(type) %>% summarise(num_ex=sum(count)) %>% mutate(isoform_type = 'novel_exon') %>% select(isoform_type, type, num_ex)
# bind_rows(event_counts, event_counts %>% )
# sum(event_counts$num_ex)
# length(omit_exons)
# # ggplot(event_counts) + 
# #   geom_col(aes(x=type, y=count, fill=type)) + 
# #   theme_minimal()
# 
# kable(event_counts)
```
|        Novel isoforms occur due to a novel arrangement of known exons or contain a novel exon which in turn is either a modification of a known exon to generate a novel modified exon, or an addition of a previously unannotated exon, or fully novel exon. For both classes of novel exon, we further annotate them with the putative biological process driving their inclusion:  alternative splicing, alternative promoter usage, or alternative polyadenylation. We find that the majority of novel exons with our dataset are novel first and last exons. 

## *de novo* transcriptomes improve sample mapping rates 

```{r, fig.width=8}

load('/Volumes/data/ocular_transcriptomes_paper/clean_data/rdata/all_tx_mappingrates.Rdata')
#***^deleteme^*** after I run the snakemake pipeline again

# df <- sample_mapping_rates %>% 
#   left_join(subtissue_to_bodyloc, .) %>% 
#   group_by(body_location) %>% 
#   summarise(mean_med_diff=mean(med_diff)*100, mean_raw_tx_count=mean(raw), mean_final_tx_count=mean(final)) %>% 
#   arrange(mean_med_diff) %>% 
#   mutate(body_location=case_when(body_location == 'Body' ~ 'Body(avg)', 
#                                  body_location == 'Brain'~ 'Brain(avg)' , 
#                                  TRUE ~ body_location)) %>% filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% mutate(body_location_pretty=gsub('_|\\.', ' ', body_location))
#   
# ggplot(data=df) + 
#   geom_col(aes(x=factor(body_location,levels = body_location,  ordered = T),y=mean_med_diff, fill=body_location)) + 
#   scale_fill_manual(values = color_list) +
#   xlab('subtissue') + 
#   ylab('median change in Salmon mapping rate') +
#   ggtitle('Changes in mapping rate between tissue specific *de novo* \ntranscriptomes and gencode V28 quantification ')+
#   coord_flip() + 
#   theme_minimal()
  
all_sample_mapping_rates <- all_sample_mapping_rates %>% 
  filter(!body_location %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line')) %>% 
  mutate(body_location_pretty=gsub('_|\\.', ' ', body_location),                                                              body_location=case_when(body_location == 'Body' ~ 'Body(avg)', 
                                 body_location == 'Brain'~ 'Brain(avg)' , 
                                  TRUE ~ body_location))

  
ggplot(data = all_sample_mapping_rates)  + 
  geom_boxplot(aes(x=build, y=mapping_rate, color=body_location)) + 
  scale_color_manual(values = color_list) +
  facet_wrap(~body_location, nrow = 3, ) +
  ylab('Salmon Mapping Rate') + 
  xlab('Transcriptome build')

```

::: {custom-style="CustomCaption"}
Figure 3 Salmon pseudo-mapping rates using both tissue specific *de novo* transcriptome and gencode transcriptomes as the underlying reference for salmon. 
:::

|        As an initial benchmark of the accuracy and utility of our transcriptome builds, we quantified transcript expression of our samples using the alignment free quantification tool salmon. Each sample was quantified twice, once using the gencode V27 human transcript annotation, and once using its associated tissue specific transcriptome. We found that globally, salmon mapping rates increased when using the *de novo* transcriptomes to as the reference transcriptome for quantification, despite average of 3 fold reduction in annotation size. Using our transcriptomes, we quantified a total of 13420034 reads across all of our samples. Additionally, for most tissue types using *de novo* transcriptomes led to a more precise estimation of gene expression.


## Novel Isoforms in Ocular tissues 

```{r, fig.width=10, fig.height=10}
load('clean_data/rdata/novel_isoforms.Rdata')
source('~/scripts/R_raincloud.R')
plot_list <- novel_eye_tx_by_tissue[!names(novel_eye_tx_by_tissue) %in% c('Lens_Stem.Cell.Line', 'ESC_Stem.Cell.Line') ]
names(plot_list) <- gsub('_|\\.', ' ', names(plot_list))
png('/tmp/upset.png', res=100, height = 1000, width = 2000)
upset(fromList(plot_list),nintersects = 20, nsets = length(novel_eye_tx_by_tissue), order.by = 'freq')
dev <- dev.off()

us <- as_ggplot(rasterGrob(readPNG('/tmp/upset.png'), interpolate = T))

### PIU
# p <- ggviolin(piu_df, x='stage', y='piu', #fill = 'tissue', 
#               trim=T, draw_quantiles = T,
#                #title ='Comparison of percent isoform usage(piu) of novel \ntranscripts in fetal and adult eye tissues', 
#               ylab = 'fraction of total gene expression')+
#     #stat_compare_means(label.y = 1.1) +
#     scale_fill_manual(values = c(Cornea='green', Retina='blue', RPE='red'))
# piu <- facet(p,facet.by = 'tissue') +
#     theme_minimal()
# piu
 #RAIN CLOUD BITCHES
piu_df <- piu_df %>% mutate(new_stage=ifelse(stage == 'adult', 'Adult', 'Fetal'), 
                            subtissue=paste0(tissue, '_', new_stage, '.Tissue'))
piu <- ggplot(data=piu_df, aes(x=stage, y=piu, fill=subtissue, color=subtissue)) + 
  geom_flat_violin(position = position_nudge(x = .2, y = 0),adjust = 2)+
  geom_point(position = position_jitter(width = .15), size = .25, alpha=.2) + 
  geom_boxplot(outlier.shape = NA, alpha = 1, width = .1, colour = "BLACK")+ 
  scale_fill_manual(values = color_list) +
  scale_color_manual(values = color_list) +
  ylab('piu')+xlab('tissue')+coord_flip()+ theme_minimal()+ facet_wrap(~tissue, nrow = 3)

### exon locations
locs <- ggplot(location_df) + 
    geom_bar(aes(x=age, fill=location, y=count), position = 'fill', stat = 'identity' ) +
    scale_fill_manual(values = c('yellow', 'purple', 'orange')) +
    facet_wrap(~ tissue) +  
    ylab('percentage of novel exons') #+ 
    #ggtitle('location of novel exons in occular tissues')

bottom <- ggarrange(piu, locs, nrow = , ncol = 2, labels = c('B', 'C'))

ggarrange(us, bottom, nrow = 2, ncol = 1, labels = 'A' )

```

::: {custom-style="CustomCaption"}
Figure 4. Analysis of novel isoform chaacteristics
  A. Set intersection of novel isoforms in ocular transcriptomes
  B. Raincloid plots of novel isoform piu. Boxplots are overlaid over piu data with estimated distribution of data set above each boxplot
  C. Location of Novel exons with in the body of associated parent transcript. CDS refers to coding region of transcript, UTR to untranslated region and NC to noncoding transcript
:::

|        Next, we analyzed the novel isoforms within our ocular transcriptomes. We compared the overlap in constructed novel isoforms across ocular tissues and found that 95 % of novel isoforms are specific to a singular ocular subtissue. (fig 3a) We then calculated for each novel isoform percent isoform usage (PIU), or the fraction of total gene expression a transcript contributed to its parent gene.  We found that on average for Retina and RPE, novel isoforms contribute to 30 percent of their parent gene’s expression, whereas in  the Cornea  we saw that novel isoforms contributed on average to 75%  their parent gene’s expression.
|        We next found the longest open reading frame for each novel isoform with a novel exon in order to see if the novel exon in the novel isoform caused a change in the protein coding region of the transcript. Novel isoforms with no detectable open reading frame had associated novel exons marked as noncoding. We found that novel exons lie largely in the untranslated regions of the transcript body. Less than 10% of detected novel isoforms potentially cause a change in its translated protein. 


## A companion visualization tool enables easy use of *de novo* transcriptomes 

![](/Volumes/data/ocular_transcriptomes_paper/clean_data/shiny_app_screenshot.png)

::: {custom-style="CustomCaption"}
Figure 5. Screenshots from *de novo* transcriptome visualization tool. From Top to bottom: dynamic PIU bar plot for selected gene and tissue. Exon level diagram of transcript body. graph of fractions of Samples transcripts were detecgted in. 
:::

|        To make our results easily accessible, we designed a webapp for visualizing and accessing our *de novo* transcriptomes. Users start by selecting gene or search for a gene by genomic location, and selecting up to 5 tissues to visualize transcript expression in. For each tissue we show the PIU for each transcript associated with a gene. We show the exon-intron structure of each transcript and mousing over exons show genomic location overlapping SNPs, and phylogentic conservation score. We additionally show a barplot of the fraction of samples in each tissue each transcript was constructed in. Users can also download the *de novo* transcriptomes for selected tissues in GTF format. 