---
title: "de novo Occular transcriptomes: results"
output: html_notebook 
---

```{r setup, echo=F}
#knitr::opts_knit$set(root.dir = '~/NIH/occular_transcriptomes_paper/')
#knitr::opts_knit$set(root.dir = '/Volumes/data/occular_transcriptomes_paper/')
knitr::opts_knit$set(root.dir = '/data/swamyvs/occular_transcriptomes_paper')
knitr::opts_chunk$set(echo = F, message = F, warning = F)
```



```{r prep}
library(tidyverse)
library(DT)
```


After our trancriptome, we reconstructed much of the refernece transcriptome, as well as novel transcripts, which we define as novel isoforms of known genes, and novel loci, which we define as transcripts whose complete sequence has not been previously annotated

```{r overall_stats, message=FALSE, warning=FALSE}
load('clean_data/overall_stats.Rdata')
overall_stats %>% datatable
```

Novel transcripts are broken down into 3 catagories - noncoding, coding but no change in the coding region of the transcript, or a change in the ocidng region of the transcript and novel loci are described as  protein coding or non coding. 
```{r tx_by_tissue, fig.height= 3.5, fig.width=10 }
load('clean_data/novel_tx_by_tissue.Rdata')
ggplot(data = novel_transcripts_by_tissue, aes(x=tissue, y=count, fill=type))+
  geom_bar(stat = 'identity',position = 'dodge') + 
  ylab(label = 'number of transcripts')+
  ggtitle(label = 'Constructed Novel Transcripts by Tissue')+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
ggplot(data = novel_loci_by_tissue, aes(x=tissue, y=count, fill=type))+
  geom_bar(stat = 'identity', position = 'dodge') + 
  ylab(label = 'number of novel loci')+
  ggtitle(label = 'Constructed Novel loci by Tissue')+
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

```

Novel transcripts are novel due to  one or more novel/unannotated exon. We can  group novel exons into 2 distinct categories: modifcations of existing exons - alternative exons, or completely distinct regions of transcibed sequence - fully novel exons. We then wanted to identify what biological process might be driving the formation of these novel exons
3 major sources of transcriptional variation are alternative splicing, Alternative first exon usage, and alternative terminal exon usage(alternative polyadenylation).


```{r classification_of_novel_exons, fig.width=7}
load('clean_data/exon_classification.Rdata')
total_exons_desc %>% datatable
rna_proc_type %>% mutate(count=as.numeric(count)) %>% datatable
```



Next, using the contructed transcriptomes for each tisssue, we analyzed splicing patterns for both reference and novel exons 


```{r splicing_heatmap, fig.height=4, fig.width=8}
library(ComplexHeatmap)
library(viridis)
load('clean_data/splicing_heatmap.Rdata')
splicing_heatmap
```

While the expression of some exons is ubiquitous across tissues, we can see clear regions in the heatmap that denote tissue specifc splicing



To illustrate the utility of our denovo transcriptomes, we reanalyzed a data set...


```{r}


```



Overlap variants onto novel exons 


```{r}

```



Finally we provide a tool to visual our results via an online webapp

```{r}

```









